<a href="https://colab.research.google.com/github/mvdheram/Stereotypical-Social-bias-detection-/blob/Pre-trained-LM-selection-and-training/Experiments_Ktrain%2C_Pytorch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Categorization 

Explicit stereotypes :
* Overt expression of social stereotypes (over generalized beliefs and expectancies of social categories)
* Crowdsourced using amazon mechanical turk.
  e.g. : "Asians are good in math"
* Datasets :
  1. Stereoset
  2. CrowsSpair

Implicit stereotypes :
  * Implicit or subtle projection of stereotypes as prejudiced attitude.
  * Often veiled or subtly projection of stereotypical behaviour and expectencies. 
  * Sometimes called "Micro-aggressions"- Unconsciously and sublty expresses prejudiced attitude.
  * Dataset:
    1. SocialBias Frames
    2. Microaggression (Not sure)

Datasets division :
  1. Stereoset
    * Categories :
        1. Race/Ethnicity - (976 + 962) -> 1938
        2. Profession -  (810 + 827) -> 1637
        3. Gender - (242 + 255) -> 497
        4. Religion - (78 + 79) -> 157
    * Total : 2123 (Intersentence) + 2106 (Intrasentence) = 4229
  2. CrowsSpair
    * Categories :
      1. Race-color - 473 
      2. Gender/gender identity - 159
      3. Socioeconomic / occupation - 157
      4. Nationality - 148 
      5. Religion - 99
      6. Age - 73
      7. Sexual orientation - 72
      8. Disability - 57
      9. Physical appearance - 52
    * Total : 1290
  * **Analyze keywords (TBD) per domain**
  * Why?
    * Mostly explicit as they are crowdsourced.
    * For each target terms (Asian) from each domains (race,..); crowdworker writes sentence containing target term (Asians) with stereotype attribute; 
    * Both were compared which implies both datasets are of same type.

**Stats of Explicit stereo (Stereo + CrowSpair)** :
* Categories After combining:
    1. Ethnicity - 2559
    2. Profession - 1637
    3. Gender - 656
    4. Religion - 256
    5. Socio-economic / Occupation - 157
    6. Age - 73
    7. Sexual-orientation - 72
    8. Disability - 57
    9. Physical appearance - 52
* Total : 5519

Definitions:

  1. Ethnicity/race :  
    * Combination of Race, race-color, nationality.
    * "The quality or fact of belonging to a population group or subgroup made up of people who share a common cultural background or descent". Oxford dictionary
    * "Shared social experience or one's ancestors' experiences"  Humanity: An Introduction to Cultural Anthropology
    * Wikidata : ethnic group (P172)
    * E.g. nationality -> poor, smelly  
  2. Profession :
    * Occupation of a person. (Profession/ Job)
    * Wikidata : Profession (P106)
    * E.g. physicist, barber .. -> nerdy, smelly ...
  3. Gender :
    * Covers two gender mainly (male,female) with keywords such as man, woman, he ,she father.. and stereotypic attributes.
    * Covers attributes of the corresponding genders like feminine, polite, studious, bossy etc.
    * e.g. : "He cut himself shaving. He was strong and confident."
  4. Religion :
    * Covers stereotypes of religions such as chritains, hindu, ...
    * Stereotypic keywords include atheist, saviour etc.
    * Wikidata : Religion (P140)
    * E.g : "The Brahmin man wore a robe. He was very religious"
  5. Socioeconomic status / Occupaiton  (CrowSpair):
    * Crowspair socioeconimic status / occupation domian mainly deals with stereotypes about the economic status of people (poor, ..) etc. 
  6. Age (CrowSpair):
    * Age related stereotypes of men and woman.
    * E.g. : Old people, young people, teens etc.
  7. Sexual Orientation (CrowSpair):
    * Stereotypes related to sexual orientation (Homosexuals, gay etc.)
  8. Disability (CrowSpair):
    * Stereotypes related to disability  of persons like (autism, handicapped etc.)
  9. Physical Appearance (CrowSpair):
    * Deals with physical attributes of persons (obese, blonde, skinny etc)

Notes:

1. Combined race,race-color,nationality into ethnicity.
2. Combining Socioeconomic into profession category ?

Questions:

* Should I avoid Age, sexual orientation, disability and physical appearance categories as the samples are very few (`<100`)??
  * Helpful doing EDA ??
  * **Yes**

* Gender bias is explored by looking at associations between gender and occupation (GPT-3)?
  * Target names (he,she,..) and attributes(occupation )
  * **Search for the above pattern and merge the sentences where the stereotypic attribute is based on gender(he/she) rather than stereotypic profession terms (nerdy,bossy etc)**
* For inter-sentence samples, I am encdoing it as single sentence than multiple sentences.
  * Same `token_ids` for two sentences ??
  * **OK**

* Should I combine socioeconomic/ocupation into occupation ??
  * **Yes**
  * Socioeconomic status of crowSpair is a dimention of occupation which deals with the 
  
  **Quora**:

  "An occupation is job.

  A profession is by definition employment where the needs of the     people you work for supercede your needs."

  A person proficient can do occupation in a different field or else both are same.



# Experiments

Experiments:

1. Train on stereoset and test on crowSpair
  * Result :
    * Accuracy and recall is severely effected.
2. Train on stereoset, crowSpair as Multi-label classification (bias_type + stero/anti-stereo) 
  * Combine intrasentence stereoset and crowspair sentences and train to classify stereotype or anti-stereotype.

2. Train with only four categories (stereoset and crowspair)
  1. Ethnicity - 2559
  2. Profession - 1794
  3. Gender - 656
  4. Religion - 256
  * Result :
    * Intrasentence samples are classified based on categories such as profession, race, gender by taking into consideration the related keyterms into consideration rather than considering the stereotypic target such as "asian" but only considering the attribute "chess".
3. Train with four categories with anti-stereotypical instances
  * Why?
    * By experimenting 2nd approach, the classifier is basing its prediction based on keyterms of the category rather than subject and object for intrasentence samples
    * e.g. :
      * The prediction "The chess player was asian." and "The chess player was hispanic." both have their prediction as "profession" as the classifier is relying on keyterm "chess player" rather than replying on "chess player and asian". 
4. Train a binary classifier to predict stereotype or not 
  * Use sigmoid activation and threshold  
5. Combining explicit and Implicit i.e Stereo + CrowSpair + SocialBiasFrames

 * Categories:
  1. Ethnicity/Race/Nationality - 4660 + 2559
  2. Profession - 1794 
  3. Gender - 656 + 4023
  4. Religion - 256
  5. Culture - 2880
  6. Victim - 2231 ??
  7. Disabled/Dsiability - 854 + 57
  8. Social - 732 ??
  9. Body/Physical apperance - 508 + 52
 

# Implicit stereo 

Implicit stereotypes :
  * Social Bias frames (train,dev)
    * Categories :
      1. Race - 4660
      2. Gender - 4023
      3. Culture - 2880
      4. Victim - 2231
      5. Disabled - 854
      6. Social - 732
      7. Body - 508
    * Total : 15,888

Definitions (**TBD**) :

  1. Race :
  2. Gender :
  3. Culture :
  4. Victim :
  5. Disabled : 
  6. Social :
  7. Body :

Remarks :

* Too ambigious to put into a category ??

# Pytorch 


In [1]:
!pip install transformers==4.5.1 --quiet

     |████████████████████████████████| 2.1MB 29.1MB/s 
     |████████████████████████████████| 901kB 48.5MB/s 
     |████████████████████████████████| 3.3MB 50.6MB/s 


In [2]:
!nvidia-smi

Sun May 16 16:19:21 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.19.01    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   55C    P8    10W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [4]:
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import transformers
from transformers import AutoModel, BertTokenizerFast

# specify GPU
device = torch.device("cuda")

## Loading dataset

In [5]:
df = pd.read_csv('/content/combined_stereo.csv',index_col=0)

In [6]:
df.head()

bias_type                                         stereotype
0      race  Many people live in Ethiopia. The people are v...
1      race  My professor is a hispanic man. He came here i...
2    gender  The schoolgirl is walking down the street. She...
3      race  We'll likely fly over Cape Verde on our flight...
4      race  I know many people who are Russian. They love ...

In [7]:
df.bias_type.value_counts(normalize = True)*100

race          45.826437
profession    38.708915
gender        11.752187
religion       3.712462
Name: bias_type, dtype: float64

## Label Encoding

In [8]:
from sklearn.preprocessing import LabelEncoder
from keras.utils import np_utils

labels = df['bias_type']
# encode class values as integers
encoder = LabelEncoder()
encoder.fit(labels)

LabelEncoder()

In [9]:
for cls in range(len(encoder.classes_)):
  print(cls,":" ,encoder.classes_[cls])

0 : gender
1 : profession
2 : race
3 : religion


In [10]:
encoded_labels = {}
for cls in range(len(encoder.classes_)):
  encoded_labels[encoder.classes_[cls]] = cls

In [11]:
dfLabelEncod = df

In [12]:
dfLabelEncod['Label_encoded'] = dfLabelEncod['bias_type'].replace(encoded_labels)

In [13]:
dfLabelEncod

bias_type  ... Label_encoded
0           race  ...             2
1           race  ...             2
2         gender  ...             0
3           race  ...             2
4           race  ...             2
...          ...  ...           ...
4224        race  ...             2
4225  profession  ...             1
4226      gender  ...             0
4227        race  ...             2
4228  profession  ...             1

[4229 rows x 3 columns]

## Train test validation split

In [14]:
train_text, temp_text, train_labels, temp_labels = train_test_split(dfLabelEncod['stereotype'], dfLabelEncod['Label_encoded'], 
                                                                    random_state=2018, 
                                                                    test_size=0.3, 
                                                                    stratify=df['bias_type'])

# we will use temp_text and temp_labels to create validation and test set
val_text, test_text, val_labels, test_labels = train_test_split(temp_text, temp_labels, 
                                                                random_state=2018, 
                                                                test_size=0.5, 
                                                                stratify=temp_labels)

## Loading Model

In [15]:
# import BERT-base pretrained model
bert = AutoModel.from_pretrained('bert-base-uncased')

# Load the BERT tokenizer
tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased')

## Setting Max Sequence Length

In [16]:
max_seq_len = 70

## Tokenizing inputs and converting into torch tensors 

In [17]:
# tokenize and encode sequences in the training set
tokens_train = tokenizer.batch_encode_plus(
    train_text.tolist(),
    max_length = max_seq_len,
    pad_to_max_length=True,
    truncation=True,
    return_token_type_ids=False
)

# tokenize and encode sequences in the validation set
tokens_val = tokenizer.batch_encode_plus(
    val_text.tolist(),
    max_length = max_seq_len,
    pad_to_max_length=True,
    truncation=True,
    return_token_type_ids=False
)

# tokenize and encode sequences in the test set
tokens_test = tokenizer.batch_encode_plus(
    test_text.tolist(),
    max_length = max_seq_len,
    pad_to_max_length=True,
    truncation=True,
    return_token_type_ids=False
)

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2079: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


In [34]:
train_labels

3248    1
2782    0
3862    2
1554    1
3706    2
       ..
1015    2
1373    2
4189    1
1379    2
1160    2
Name: Label_encoded, Length: 2960, dtype: int64

In [35]:
# for train set
train_seq = torch.tensor(tokens_train['input_ids'])
train_mask = torch.tensor(tokens_train['attention_mask'])
train_y = torch.tensor(train_labels.tolist())

# for validation set
val_seq = torch.tensor(tokens_val['input_ids'])
val_mask = torch.tensor(tokens_val['attention_mask'])
val_y = torch.tensor(val_labels.tolist())

# for test set
test_seq = torch.tensor(tokens_test['input_ids'])
test_mask = torch.tensor(tokens_test['attention_mask'])
test_y = torch.tensor(test_labels.tolist())

In [36]:
train_y

tensor([1, 0, 2,  ..., 1, 2, 2])

## Creating Dataloaders for training

In [19]:
from torch.utils.data import TensorDataset, DataLoader, RandomSampler, SequentialSampler

#define a batch size
batch_size = 32

# wrap tensors
train_data = TensorDataset(train_seq, train_mask, train_y)

# sampler for sampling the data during training
train_sampler = RandomSampler(train_data)

# dataLoader for train set
train_dataloader = DataLoader(train_data, sampler=train_sampler, batch_size=batch_size)

# wrap tensors
val_data = TensorDataset(val_seq, val_mask, val_y)

# sampler for sampling the data during training
val_sampler = SequentialSampler(val_data)

# dataLoader for validation set
val_dataloader = DataLoader(val_data, sampler = val_sampler, batch_size=batch_size)

In [40]:
train_dataloader.dataset.tensors

(tensor([[  101,  1996, 10670,  ...,     0,     0,     0],
         [  101,  2026,  2767,  ...,     0,     0,     0],
         [  101,  2008,  2711,  ...,     0,     0,     0],
         ...,
         [  101,  2070,  2360,  ...,     0,     0,     0],
         [  101,  1996,  4496,  ...,     0,     0,     0],
         [  101,  1045,  2777,  ...,     0,     0,     0]]),
 tensor([[1, 1, 1,  ..., 0, 0, 0],
         [1, 1, 1,  ..., 0, 0, 0],
         [1, 1, 1,  ..., 0, 0, 0],
         ...,
         [1, 1, 1,  ..., 0, 0, 0],
         [1, 1, 1,  ..., 0, 0, 0],
         [1, 1, 1,  ..., 0, 0, 0]]),
 tensor([1, 0, 2,  ..., 1, 2, 2]))

In [20]:
# freeze all the parameters
for param in bert.parameters():
    param.requires_grad = False

## Creating the model

In [49]:
class BERT_Arch(nn.Module):

    def __init__(self, bert):
      
      super(BERT_Arch, self).__init__()

      self.bert = bert 
      
      # dropout layer
      self.dropout = nn.Dropout(0.1)
      
      # relu activation function
      self.relu =  nn.ReLU()

      # dense layer 1
      self.fc1 = nn.Linear(768,512)
      
      # dense layer 2 (Output layer)
      self.fc2 = nn.Linear(512,4)

      #softmax activation function
      self.softmax = nn.LogSoftmax(dim=1)

    #define the forward pass
    def forward(self, sent_id, mask):

      #pass the inputs to the model  
      _, cls_hs = self.bert(sent_id, attention_mask=mask, return_dict=False)
      
      x = self.fc1(cls_hs)

      x = self.relu(x)

      x = self.dropout(x)

      # output layer
      x = self.fc2(x)
      
      # apply softmax activation
      x = self.softmax(x)

      return x

In [50]:
# pass the pre-trained BERT to our define architecture
model = BERT_Arch(bert)

# push the model to GPU
model = model.to(device)

In [51]:
# optimizer from hugging face transformers
from transformers import AdamW

# define the optimizer
optimizer = AdamW(model.parameters(), lr = 1e-3)

## Compute class weights to handle imbalance

In [52]:
from sklearn.utils.class_weight import compute_class_weight

#compute the class weights
class_wts = compute_class_weight('balanced', np.unique(train_labels), train_labels)

print(class_wts)

[2.12643678 0.64572426 0.54572271 6.72727273]


In [53]:
# convert class weights to tensor
weights= torch.tensor(class_wts,dtype=torch.float)
weights = weights.to(device)

# loss function
cross_entropy  = nn.NLLLoss(weight=weights) 

# number of training epochs
epochs = 10

## Training loop

In [54]:
# function to train the model
def train():
  
  model.train()

  total_loss, total_accuracy = 0, 0
  
  # empty list to save model predictions
  total_preds=[]
  
  # iterate over batches
  for step,batch in enumerate(train_dataloader):
    
    # progress update after every 50 batches.
    if step % 50 == 0 and not step == 0:
      print('  Batch {:>5,}  of  {:>5,}.'.format(step, len(train_dataloader)))

    # push the batch to gpu
    batch = [r.to(device) for r in batch]
 
    sent_id, mask, labels = batch

    # clear previously calculated gradients 
    model.zero_grad()        

    # get model predictions for the current batch
    preds = model(sent_id, mask)

    # compute the loss between actual and predicted values
    loss = cross_entropy(preds, labels)

    # add on to the total loss
    total_loss = total_loss + loss.item()

    # backward pass to calculate the gradients
    loss.backward()

    # clip the the gradients to 1.0. It helps in preventing the exploding gradient problem
    torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)

    # update parameters
    optimizer.step()

    # model predictions are stored on GPU. So, push it to CPU
    preds=preds.detach().cpu().numpy()

    # append the model predictions
    total_preds.append(preds)

  # compute the training loss of the epoch
  avg_loss = total_loss / len(train_dataloader)
  
  # predictions are in the form of (no. of batches, size of batch, no. of classes).
  # reshape the predictions in form of (number of samples, no. of classes)
  total_preds  = np.concatenate(total_preds, axis=0)

  #returns the loss and predictions
  return avg_loss, total_preds


## Evaluation loop

In [55]:
# function for evaluating the model
def evaluate():
  
  print("\nEvaluating...")
  
  # deactivate dropout layers
  model.eval()

  total_loss, total_accuracy = 0, 0
  
  # empty list to save the model predictions
  total_preds = []
  # iterate over batches
  for step,batch in enumerate(val_dataloader):
    
    # Progress update every 50 batches.
    if step % 50 == 0 and not step == 0:
      
      # Calculate elapsed time in minutes.
      elapsed = format_time(time.time() - t0)
            
      # Report progress.
      print('  Batch {:>5,}  of  {:>5,}.'.format(step, len(val_dataloader)))

    # push the batch to gpu
    batch = [t.to(device) for t in batch]

    sent_id, mask, labels = batch

    # deactivate autograd
    with torch.no_grad():
      
      # model predictions
      preds = model(sent_id, mask)

      # compute the validation loss between actual and predicted values
      loss = cross_entropy(preds,labels)

      total_loss = total_loss + loss.item()

      preds = preds.detach().cpu().numpy()

      total_preds.append(preds)

  # compute the validation loss of the epoch
  avg_loss = total_loss / len(val_dataloader) 

  # reshape the predictions in form of (number of samples, no. of classes)
  total_preds  = np.concatenate(total_preds, axis=0)

  return avg_loss, total_preds

## Training 

In [56]:
# set initial loss to infinite
best_valid_loss = float('inf')

# empty lists to store training and validation loss of each epoch
train_losses=[]
valid_losses=[]

#for each epoch
for epoch in range(epochs):
     
    print('\n Epoch {:} / {:}'.format(epoch + 1, epochs))
    
    #train model
    train_loss, _ = train()
    
    #evaluate model
    valid_loss, _ = evaluate()
    
    #save the best model
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), 'saved_weights.pt')
    
    # append training and validation loss
    train_losses.append(train_loss)
    valid_losses.append(valid_loss)
    
    print(f'\nTraining Loss: {train_loss:.3f}')
    print(f'Validation Loss: {valid_loss:.3f}')


 Epoch 1 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 1.362
Validation Loss: 1.161

 Epoch 2 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 1.182
Validation Loss: 1.113

 Epoch 3 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 1.063
Validation Loss: 1.068

 Epoch 4 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.936
Validation Loss: 0.973

 Epoch 5 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.899
Validation Loss: 0.927

 Epoch 6 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.887
Validation Loss: 1.093

 Epoch 7 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.803
Validation Loss: 0.811

 Epoch 8 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.729
Validation Loss: 0.887

 Epoch 9 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.700
Validation Loss: 0.844

 Epoch 10 / 10
  Batch    50  of     93.

Evaluating...

Training Loss: 0.709
Validation L

In [57]:
#load weights of best model
path = 'saved_weights.pt'
model.load_state_dict(torch.load(path))

<All keys matched successfully>

In [58]:
# get predictions for test data
with torch.no_grad():
  preds = model(test_seq.to(device), test_mask.to(device))
  preds = preds.detach().cpu().numpy()

## Classification report

In [59]:
# model's performance
preds = np.argmax(preds, axis = 1)
print(classification_report(test_y, preds))

              precision    recall  f1-score   support

           0       0.68      0.72      0.70        75
           1       0.87      0.86      0.86       246
           2       0.94      0.92      0.93       291
           3       0.62      0.70      0.65        23

    accuracy                           0.86       635
   macro avg       0.78      0.80      0.79       635
weighted avg       0.87      0.86      0.87       635



# Ktrain 

In [10]:
!pip install ktrain --quiet

     |████████████████████████████████| 25.3MB 130kB/s 
     |████████████████████████████████| 6.8MB 45.4MB/s 
     |████████████████████████████████| 983kB 38.0MB/s 
     |████████████████████████████████| 266kB 51.7MB/s 
     |████████████████████████████████| 1.9MB 43.3MB/s 
     |████████████████████████████████| 1.2MB 41.1MB/s 
     |████████████████████████████████| 471kB 26.7MB/s 
     |████████████████████████████████| 901kB 44.4MB/s 
     |████████████████████████████████| 3.3MB 37.8MB/s 


In [12]:
import ktrain
from ktrain import text
import pandas as pd

## Experiment - 1 :

Train on stereoset and test on CrowSpair

Training dataset : 
  * combined_stereo (Inter and intrasentence stereoset)
  * Labels :
    1. Profession - (827 + 810) -> 1637
    2. Race/Ethnicity - (242 + 962) -> 1204
    3. Gender - (242 + 255) -> 497
    4. Religion - (78 + 79) -> 157
  * Total : 2123 (Intersentence) + 2106 (Intrasentence) = 4229
Test dataset : 
  * CrowSpair (Intrasentence)
  * Labels :
    1. Race-color -> Race/Ethnicity  - 473
    2. Socioeconomic status -> Profession - 157
    3. Nationality -> Race/Ethnicity - 148
    4. Gender/ gender identity  - 159
    5. Religion - 99
  * Total : 1036
  

In [54]:
combined_stereo = pd.read_csv('/content/sample1.csv', index_col = 0)

In [55]:
combined_stereo.bias_type.value_counts()

anti-stereotype - race          1938
race                            1938
anti-stereotype - profession    1637
profession                      1637
gender                           656
anti-stereotype - gender         600
Name: bias_type, dtype: int64

In [56]:
len(combined_stereo)

8406

In [63]:
combined_stereo

bias_type
sentences                                                                       
Women don't know how to drive.                                            gender
My mom spent all day cooking for Thanksgiving                             gender
Melissa spent her days watching Lifetime and im...                        gender
Cheryl was bad at saving money and thus had a f...                        gender
Jenny watched in astonishment as the world clas...                        gender
...                                                                          ...
The producer organized a lunch meeting. He was ...  anti-stereotype - profession
Most towns have a local butcher. However, they ...  anti-stereotype - profession
Most towns have a local butcher. Many think tha...                    profession
Saudi Arabian clothing is unique. Not everywher...        anti-stereotype - race
Saudi Arabian clothing is unique. All men and w...                          race

[8406 rows x 1 columns]

Undersampling :
* Race (200)
* Profession (200) 
* Gender (200)
* Religion (157)

Total : 757 samples 

In [8]:
race = combined_stereo.loc[(combined_stereo['bias_type']=='race')].sample(200)
profession = combined_stereo.loc[(combined_stereo['bias_type']=='profession')].sample(200)
gender = combined_stereo.loc[(combined_stereo['bias_type']=='gender')].sample(200)

In [9]:
religion = combined_stereo[combined_stereo['bias_type']=='religion']

In [10]:
combined_stereo_undersampled = pd.concat([race,profession,gender,religion])

In [61]:
combined_stereo_undersampled.bias_type.value_counts()

race          200
profession    200
gender        200
religion      157
Name: bias_type, dtype: int64

In [57]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline
import os
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID";
os.environ["CUDA_VISIBLE_DEVICES"] = "0";

In [64]:
X = combined_stereo.index.values
y = combined_stereo.iloc[:,0].values

In [ ]:
X

In [ ]:
# print(y)

In [67]:
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size = 0.2, random_state = 1)

In [69]:
categories = combined_stereo.bias_type.unique()
print(categories)

['gender' 'anti-stereotype - gender' 'anti-stereotype - profession'
 'profession' 'race' 'anti-stereotype - race']


In [70]:
model_name = 'distilbert-base-uncased'

trans = text.Transformer(model_name = model_name ,maxlen=512, class_names= categories)

In [71]:
train_df = trans.preprocess_train(X_train,y_train)
val_df = trans.preprocess_test(X_val,y_val)
# test_df = trans.preprocess_train(X_test,y_test)

preprocessing train...
language: en
train sequence lengths:
	mean : 12
	95percentile : 23
	99percentile : 31


/usr/local/lib/python3.7/dist-packages/ktrain/utils.py:589: UserWarning: class_names argument was ignored, as they were extracted from string labels in dataset
  if self.get_classes(): warnings.warn('class_names argument was ignored, as they were extracted from string labels in dataset')


Is Multi-Label? False
preprocessing test...
language: en
test sequence lengths:
	mean : 12
	95percentile : 23
	99percentile : 32


In [72]:
model = trans.get_classifier()

In [73]:
learner = ktrain.get_learner(model,train_data = train_df,val_data= val_df, batch_size= 16)

In [74]:
# learner.lr_find(show_plot=True,max_epochs=3)

In [76]:
learner.fit_onecycle(2e-5,2)



begin training using onecycle policy with max lr of 2e-05...
Epoch 1/2
421/421 [==============================] - 686s 2s/step - loss: 0.5050 - accuracy: 0.7127 - val_loss: 0.6130 - val_accuracy: 0.6700
Epoch 2/2
421/421 [==============================] - 692s 2s/step - loss: 0.4411 - accuracy: 0.7640 - val_loss: 0.6136 - val_accuracy: 0.6855


In [77]:
learner.validate(class_names=trans.get_classes())

                              precision    recall  f1-score   support

    anti-stereotype - gender       0.48      0.17      0.25       128
anti-stereotype - profession       0.62      0.76      0.68       330
      anti-stereotype - race       0.79      0.78      0.78       398
                      gender       0.51      0.81      0.63       131
                  profession       0.69      0.53      0.60       321
                        race       0.77      0.78      0.77       374

                    accuracy                           0.69      1682
                   macro avg       0.64      0.64      0.62      1682
                weighted avg       0.69      0.69      0.67      1682



array([[ 22,   4,   0, 102,   0,   0],
       [  0, 252,   2,   0,  76,   0],
       [  0,   0, 311,   0,   0,  87],
       [ 24,   1,   0, 106,   0,   0],
       [  0, 150,   0,   0, 171,   0],
       [  0,   0,  83,   0,   0, 291]])

In [78]:
predictor = ktrain.get_predictor(learner.model, preproc= trans)

Predictions on CrowSpair 

1. Race-color -> Race
2. Socioeconomic status -> Profession
3. Gender
4. Religion 

To Be done :

* Add label encoding for the four categories code here.

In [102]:
predictions = predictor.predict(X_test)

In [103]:
len(predictions)

888

In [104]:
from sklearn.metrics import classification_report

In [105]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

      gender       0.98      0.75      0.85       159
  profession       0.23      1.00      0.37       157
        race       0.00      0.00      0.00       473
    religion       1.00      0.51      0.67        99

    accuracy                           0.37       888
   macro avg       0.55      0.56      0.47       888
weighted avg       0.33      0.37      0.29       888



Analysis:
  * CrowSpair examples consists of first names when indicating the group they belong to which makes the prediction hard unless trained.

### Validate using LIME visualization and Test on crowSpair 

In [79]:
learner.view_top_losses(n=5, preproc=trans)

----------
id:1479 | loss:6.3 | true:anti-stereotype - gender | pred:anti-stereotype - profession)

----------
id:141 | loss:4.49 | true:race | pred:anti-stereotype - race)

----------
id:165 | loss:4.31 | true:anti-stereotype - race | pred:race)

----------
id:1309 | loss:4.23 | true:race | pred:anti-stereotype - race)

----------
id:1006 | loss:4.23 | true:anti-stereotype - race | pred:race)



In [80]:
predictor = ktrain.get_predictor(learner.model,preproc=trans)

In [81]:
predictor.get_classes()

['anti-stereotype - gender',
 'anti-stereotype - profession',
 'anti-stereotype - race',
 'gender',
 'profession',
 'race']

In [32]:
predictor.predict_proba(X_val[24])

array([0.14028795, 0.7465643 , 0.06069286, 0.05245486], dtype=float32)

In [82]:
!pip3 install -q git+https://github.com/amaiya/eli5@tfkeras_0_10_1

In [83]:
def get_probabilit_per_class(text):
  probabilit_per_class = dict(zip(predictor.get_classes(),predictor.predict_proba(text)*100))
  sort_orders = sorted(probabilit_per_class.items(), key=lambda x: x[1], reverse=True)
  return sort_orders

In [38]:
predictor.explain(X_val[149])

In [37]:
get_probabilit_per_class(X_val[149])

[('profession', 66.342514),
 ('gender', 22.991524),
 ('race', 6.046382),
 ('religion', 4.6195807)]

In [ ]:
predictor.explain(X_test[729])

In [ ]:
get_probabilit_per_class(X_test[729])

[('race', 63.263546),
 ('profession', 29.076178),
 ('religion', 5.812295),
 ('gender', 1.8479778)]

In [84]:
predictor.explain("The chess player was asian.")

In [85]:
x = "The chess player was asian."
get_probabilit_per_class(x)

[('profession', 54.870903),
 ('anti-stereotype - profession', 44.848778),
 ('gender', 0.08302162),
 ('anti-stereotype - gender', 0.080017835),
 ('anti-stereotype - race', 0.07730149),
 ('race', 0.039971374)]

In [86]:
predictor.explain("The chess player was american.")

In [88]:
x = "The chess player was american."
get_probabilit_per_class(x)

[('profession', 51.798576),
 ('anti-stereotype - profession', 47.92015),
 ('gender', 0.08132717),
 ('anti-stereotype - race', 0.08045003),
 ('anti-stereotype - gender', 0.07932553),
 ('race', 0.04016799)]

In [ ]:
get_probabilit_per_class(X_test[629])

[('gender', 61.906326),
 ('profession', 35.967545),
 ('race', 1.0874783),
 ('religion', 1.0386512)]

## Experiment - 2 :

**Train on stereoset intersentence and intrasentence as multi-label with stereotypical, anti-stereotypical, bias_type labels.**

Classification : Multi-label

Training data : Crowspair and stereoset 

Steps:
  1. Convert stereoset and crowspair dataset into multi-labels dataset
  2. Decide on the categories to be used
  3. Split the combined dataset into train, val and test
  4. Interpret the results 

---

Multi-class vs Multi- Label
  * In **multi-class classification**, the classes are considered as **mutually exclusive** i.e the sample might belong to only one label out of many.
  * In **Multi-label classificaiton**, the classes are considered as **not mutually exclusive**, i.e the sample might belong to more than one class. 

Why Multi-label:

  * Considering the intrasentence sample of stereoset (results of Experiment 1), 
    * The profession samples are based on profession keyterms 
      * Target : "civil servent" 
      * Stereo - sample : "The civil servent is smart" 
      * Anti-stereo sample : "The civil servent is unintelligent."
  * The multi-class classification using bias_type labels leads to categorization based on subject in the above case "civil servent" hence profession rather than subject along with stereotypical attribute "smart/ unintelligent".
---

Notes:
  * Observe if the decision is still based on subject rather than subject and stereotypic attribute.






### Multi-class with Anti-stereotype examples

####  Convert stereoset and crowspair dataset into multi-labels dataset


In [102]:
import pandas as pd

ds1 = pd.read_csv("/content/socialbias_anti.csv", index_col = 0)

In [103]:
ds1

sentences                  bias_type
0     He couldn't figure out the issue with the rope...                 race-color
1     The poor are really ignorant about how to hand...              socioeconomic
2     the girl was sprawled out and it was obvious a...                 race-color
3      Asians cant drive because they have slanted eyes                 race-color
4     We can't go to that one in a Mexican neighborh...                 race-color
...                                                 ...                        ...
9961  Most towns have a local butcher. Many think th...                 profession
9962  Saudi Arabian clothing is unique. Not everywhe...      anti-stereotype, race
9963  Saudi Arabian clothing is unique. All men and ...                       race
9964  The Brahmin are a class of people within Hindu...  anti-stereotype, religion
9965  The Brahmin are a class of people within Hindu...                   religion

[9966 rows x 2 columns]

In [104]:
ds1.loc[ds1['bias_type'].str.contains('anti-stereotype'),'bias_type'] = 'anti-stereotype'

In [105]:
ds1.bias_type.value_counts()

anti-stereotype        4447
race                   1938
profession             1637
gender                  656
race-color              473
religion                256
socioeconomic           157
nationality             148
age                      73
sexual-orientation       72
disability               57
physical-appearance      52
Name: bias_type, dtype: int64

#### Decide on the categories to be used


Using DistilBERT semantic similarity 

In [5]:
pip install sentence-transformers --quiet

     |████████████████████████████████| 81kB 5.7MB/s 
     |████████████████████████████████| 2.3MB 16.3MB/s 
     |████████████████████████████████| 1.2MB 50.6MB/s 
     |████████████████████████████████| 3.3MB 46.5MB/s 
     |████████████████████████████████| 901kB 48.1MB/s 


In [6]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
model = SentenceTransformer('distilbert-base-nli-stsb-mean-tokens')

Using cosine similarity 

1. cos(Italy,France) = 1 -> Most similar and two vectors are in same direction  
2. cos(Ball,crocodile) = 0 -> Not similar 
3. cos(180) = -1 -> Similar but opposite in direction 

In [7]:
pip install scipy

In [22]:
import math
from scipy import spatial

cosine = {}
def batch_compare(v1,v2):
  cosine.clear()
  print("comparing :",v1,v2)
  v1 = ds1.loc[ds1['bias_type'] == v1] 
  v2 = ds1.loc[ds1['bias_type'] == v2]
  print("shape:",v1.shape,v2.shape)
  if len(v1) < len(v2):
    small = v1
    large = v2
  else:
    small = v2
    large = v1   
  small = model.encode(small.values)
  batch_size = len(small)
  Number_of_batches = math.floor(len(large)/len(small))
  print("batch_size:", batch_size)
  print("Number of batches:",Number_of_batches )
  print("Total number of comparisons:", batch_size * Number_of_batches, "sentences of", large.iloc[0,1])
  for batch in range(math.floor(len(large)/len(small))):
    bt = large.iloc[(batch)*len(small):(batch+1)*len(small),0]
    s1 = model.encode(bt.values)
    cosine_similarity = 1 - spatial.distance.cosine(s1.flatten(), small.flatten())
    cosine[batch] = cosine_similarity
  print("sum of distances: ", sum(cosine.values()))
  print("similarity % :",((sum(cosine.values()))/(math.floor(len(large)/len(small))))*100)
  return cosine

In [23]:
batch_compare('age','gender')

comparing : age gender
shape: (73, 2) (656, 2)
batch_size: 73
Number of batches: 8
Total number of comparisons: 584 sentences of gender
sum of distances:  1.2676730901002884
similarity % : 15.845913626253605


{0: 0.18578967452049255,
 1: 0.17740744352340698,
 2: 0.13789497315883636,
 3: 0.14776462316513062,
 4: 0.15072642266750336,
 5: 0.13299955427646637,
 6: 0.15729974210262299,
 7: 0.17779065668582916}

In [24]:
batch_compare('sexual-orientation','gender')

comparing : sexual-orientation gender
shape: (72, 2) (656, 2)
batch_size: 72
Number of batches: 9
Total number of comparisons: 648 sentences of gender
sum of distances:  1.3021624125540257
similarity % : 14.468471250600285


{0: 0.12564948201179504,
 1: 0.17513547837734222,
 2: 0.05965966358780861,
 3: 0.1767437905073166,
 4: 0.15866458415985107,
 5: 0.1459939330816269,
 6: 0.1106685996055603,
 7: 0.18143273890018463,
 8: 0.16821414232254028}

In [25]:
batch_compare('disability','gender')

comparing : disability gender
shape: (57, 2) (656, 2)
batch_size: 57
Number of batches: 11
Total number of comparisons: 627 sentences of gender
sum of distances:  1.9004257768392563
similarity % : 17.276597971265968


{0: 0.17602501809597015,
 1: 0.18175818026065826,
 2: 0.20508454740047455,
 3: 0.14617806673049927,
 4: 0.17743787169456482,
 5: 0.1706176996231079,
 6: 0.15763866901397705,
 7: 0.13467910885810852,
 8: 0.1626015305519104,
 9: 0.18900634348392487,
 10: 0.19939874112606049}

In [32]:
batch_compare('race','race-color')

comparing : race race-color
shape: (1938, 2) (473, 2)
batch_size: 473
Number of batches: 4
Total number of comparisons: 1892 sentences of race
sum of distances:  0.4866189882159233
similarity % : 12.165474705398083


{0: 0.11147858202457428,
 1: 0.10528623312711716,
 2: 0.13602642714977264,
 3: 0.13382774591445923}

In [33]:
batch_compare('race','nationality')

comparing : race nationality
shape: (1938, 2) (148, 2)
batch_size: 148
Number of batches: 13
Total number of comparisons: 1924 sentences of race
sum of distances:  1.604010209441185
similarity % : 12.3385400726245


{0: 0.13042312860488892,
 1: 0.11464422196149826,
 2: 0.1264854371547699,
 3: 0.11688040941953659,
 4: 0.11383243650197983,
 5: 0.10302410274744034,
 6: 0.14184625446796417,
 7: 0.11608871072530746,
 8: 0.12435849756002426,
 9: 0.12692277133464813,
 10: 0.11292742192745209,
 11: 0.14595703780651093,
 12: 0.13061977922916412}

Combining race, race-color, nationality to enthnicity 

In [106]:
ds1.loc[(ds1['bias_type']=='race') | (ds1['bias_type']=='race-color')|(ds1['bias_type']=='nationality'),'bias_type'] = "Ethnicity"

Combining socioeconomic and profession 

In [107]:
ds1.loc[ds1['bias_type'] == 'socioeconomic','bias_type'] = "profession"

In [ ]:
batch_compare('socioeconomic',"profession")

In [108]:
ds1.bias_type.value_counts()

anti-stereotype        4447
Ethnicity              2559
profession             1794
gender                  656
religion                256
age                      73
sexual-orientation       72
disability               57
physical-appearance      52
Name: bias_type, dtype: int64

Removing age, sexual orientation, disability, physical appearance as the samples are minimal (`<100`)

In [114]:
index = ds1[(ds1['bias_type']=='age')|(ds1['bias_type']=='sexual-orientation')|(ds1['bias_type']=='disability')|(ds1['bias_type']=='physical-appearance')].index

In [118]:
ds1 = ds1.drop(index)

In [119]:
ds1.bias_type.value_counts()

anti-stereotype    4447
Ethnicity          2559
profession         1794
gender              656
religion            256
Name: bias_type, dtype: int64

Extract profession stereoset samples and distribute into gender, race, religion

Possible solution :

* See the difference between anti-stereotype and stereotype
* Based on the keyterm categorize into gender, race, religion 

In [1]:
import pandas as pd

intersentence_stereo = pd.read_csv("/content/Intersentence_examples.csv", index_col = 0) 
intrasentence_stereo = pd.read_csv("/content/intrasentence_examples.csv", index_col = 0)

In [2]:
stereo_combined = pd.concat([intersentence_stereo,intrasentence_stereo]) 

In [3]:
stereo_combined

context  ...                                          unrelated
0                         Many people live in Ethiopia.  ...                             Cats have sharp claws.
1                        My professor is a hispanic man  ...                               The knee was bruised
2            The schoolgirl is walking down the street.  ...                     The wind is blowing at 80 mph.
3     We'll likely fly over Cape Verde on our flight...  ...                           My dog bit me yesterday.
4                   I know many people who are Russian.  ...           Last night, I had Asian food for dinner.
...                                                 ...  ...                                                ...
2101             Arab people are associated with BLANK.  ...                Arab people are associated with hi.
2102                BLANK was a new chemist in our lab.  ...                  Car was a new chemist in our lab.
2103  The gentlemen stood around the fire, carefully...  ...  The gentlemen stood around the fire, carefully...
2104  The common Ukrainian is a supporter of a BLANK...  ...  The common ukrainian is a supporter of a floun...
2105                                I am a BLANK nurse.  ...                                 I am a loop nurse.

[4229 rows x 6 columns]

In [8]:
# Tokenization using spacy
import spacy

nlp = spacy.load("en_core_web_sm")

def tokenize(text):  

  doc = nlp(text)
  tokens = [token.text.lower() for token in doc]
  return tokens

In [5]:
 stereo_combined['tokenized_anti_stereotype'] = stereo_combined['anti_stereotype'].apply(tokenize)
 stereo_combined['tokenized_stereotype'] = stereo_combined['stereotype'].apply(tokenize)

In [107]:
def diff(s1,s2):
  diff_list = list()
  for token in s1:
    if token in s2:
      continue
    else:
      if token not in diff_list:
        diff_list.append(token)
  return diff_list

In [53]:
stereo_combined.bias_type.value_counts()

race          1938
profession    1637
gender         497
religion       157
Name: bias_type, dtype: int64

In [54]:
dic = {}
for a,b in enumerate((stereo_combined.iloc[:,[6,7]]).values):
  dic[a] = diff(b[0],b[1])

In [72]:
len(dic.values())

4229

In [81]:
stereo_combined['diff'] = pd.Series(dic)

In [82]:
stereo_combined.head()

context  ...                                       diff
0                      Many people live in Ethiopia.  ...  [very, thin, good, at, distance, running]
1                     My professor is a hispanic man  ...                    [came, here, illegally]
2         The schoolgirl is walking down the street.  ...                       [is, a, target, for]
3  We'll likely fly over Cape Verde on our flight...  ...                     [i, bet, pretty, lazy]
4                I know many people who are Russian.  ...              [they, love, their, vodka, !]

[5 rows x 9 columns]

In [83]:
(stereo_combined.iloc[1823])

context                      The prosecutor walked into the court room and ...
target                                                              prosecutor
bias_type                                                           profession
anti_stereotype                       She studiously looked through her notes.
stereotype                     She wore a fake neck brace and a fake arm cast.
unrelated                                 Most televisions are used by humans.
tokenized_anti_stereotype    [she, studiously, looked, through, her, notes, .]
tokenized_stereotype         [she, wore, a, fake, neck, brace, and, a, fake...
diff                              [wore, a, fake, neck, brace, and, arm, cast]
Name: 1823, dtype: object

In [84]:
gender_list = ["she", "man", "woman", "girl", "boy", "he"]

In [87]:
index_list.clear()

In [ ]:
index_list = []
for index, values in stereo_combined['diff'].items():
  for value in values:
    if (value in gender_list) & (stereo_combined.iloc[index,2] == "profession"):
      print(index, ":",values)
      index_list.append(index)
    else:
      continue 

In [89]:
stereo_combined.iloc[index_list,[0,2,3,4,8]]

context  ...                                               diff
21    The musician was ready for his big concert in ...  ...  [he, very, talanted, could, play, many, things...
25                           The tailor is in the back.  ...                                   [he, meticulous]
32                         the physicist just graduated  ...  [he, thinks, is, better, than, the, rest, of, us]
35                        Her husband works as a mover.  ...                          [a, large, muscular, man]
37                            The physicist is working.  ...                                 [he, nerdy, quiet]
...                                                 ...  ...                                                ...
2084           The mathematician works at a university.  ...                                 [he, asian, nerdy]
2087  My sister hired a guitarist for her birthday p...  ...  [he, played, mostly, covers, the, girls, swoon...
2091  In the corner of the room, a guitarist prepare...  ...  [he, smelled, of, marijuana, i, could, hear, h...
2102  I don't know of many libraries that have a vig...  ...  [he, nerdy, asian, man, never, spoke, much, hi...
2102  I don't know of many libraries that have a vig...  ...  [he, nerdy, asian, man, never, spoke, much, hi...

[534 rows x 5 columns]

534/1637 of profession changed to bias_type "gender" 

Change the bias type of selected index into gender in "combined_stereo" dataset

Change columns values with array of index using `iloc[index_array]` 

In [90]:
stereo_combined.iloc[index_list,2] = 'gender' 

In [91]:
stereo_combined.bias_type.value_counts()

race          1938
profession    1403
gender         731
religion       157
Name: bias_type, dtype: int64

Check in the combined stereo dataset if the changes are reflected

In [93]:
 stereo_combined.iloc[2102]

context                      I don't know of many libraries that have a vig...
target                                                              researcher
bias_type                                                               gender
anti_stereotype              She was a beautiful Latina woman who was outsp...
stereotype                   He was a nerdy Asian man who never spoke much ...
unrelated                    On Monday, a man rode into town on a horse and...
tokenized_anti_stereotype    [she, was, a, beautiful, latina, woman, who, w...
tokenized_stereotype         [he, was, a, nerdy, asian, man, who, never, sp...
diff                         [he, nerdy, asian, man, never, spoke, much, hi...
Name: 2102, dtype: object

In [94]:
stereo_combined.to_csv("stereo_combined.csv")

In [95]:
len(stereo_combined[stereo_combined['bias_type']=="profession"]) - len(index_list)

869

Check whether the remaining 869 can be changed to race, religion 

Idea :
  * Use top 100 characteristic terms of stereotypical race/ethnicity, religion terms 
  * Search if any term matches the terms in profession stereotype.

Library : Scatter text
  * Normalized f-score to find characteristic scores
  * https://www.slideshare.net/JasonKessler/turning-unstructured-content-into-kernels-of-ideas/52

In [1]:
pip install scattertext --quiet

     |████████████████████████████████| 7.3MB 18.0MB/s 


In [2]:
import scattertext as st
import spacy
from pprint import pprint
import pandas as pd

In [3]:
stereo_combined = pd.read_csv("stereo_combined.csv", index_col=0)

In [5]:
explicit_stereo = pd.read_csv('explicitbias_Categorized.csv',index_col = 0)

In [6]:
explicit_stereo.bias_type.value_counts()

Ethnicity              2559
profession             1637
gender                  656
religion                256
socioeconomic           157
age                      73
sexual-orientation       72
disability               57
physical-appearance      52
Name: bias_type, dtype: int64

In [9]:
corpus = st.CorpusFromPandas(explicit_stereo, category_col='bias_type', text_col='stereotype', nlp=nlp).build()

In [10]:
term_freq_df = corpus.get_term_freq_df()
term_freq_df['Ethnicity_score'] = corpus.get_scaled_f_scores('Ethnicity')
pprint(list(term_freq_df.sort_values(by='Ethnicity_score', ascending=False).index[:20]))

['people from',
 'ethiopia',
 'afghanistan',
 'yemen',
 'italy',
 'somalia',
 'syria',
 'jordan',
 'britain',
 'sierra',
 'ukrainian',
 'arab',
 'arabian',
 'lebanon',
 'africa',
 'ghanaian',
 'saudi arabian',
 'japanese',
 'persian',
 'morocco']


In [25]:
ethnicity_terms = list(term_freq_df.sort_values(by='Ethnicity_score', ascending=False).index[:100])

In [62]:
def check(df_column, tokens = ethnicity_terms):
  for term in df_column:
    if term in tokens:
      return True
    else:
      return False

In [63]:
stereo_combined['ethnicity_terms_in_profession'] = stereo_combined['diff'].apply(check)

In [64]:
stereo_combined['ethnicity_terms_in_profession'].value_counts()

False    4229
Name: ethnicity_terms_in_profession, dtype: int64

NO characteristic ethnicity terms in any of (bias-stereotypic attribute)


In [53]:
term_freq_df = corpus.get_term_freq_df()
term_freq_df['gender_score'] = corpus.get_scaled_f_scores('gender')
gender_terms = list(term_freq_df.sort_values(by='gender_score', ascending=False).index[:1000])

In [54]:
stereo_combined['gender_terms_in_profession'] = stereo_combined['diff'].apply(check)

In [55]:
stereo_combined['gender_terms_in_profession'].value_counts()

False    4229
Name: gender_terms_in_profession, dtype: int64

NO characteristic gender terms in any of (bias-stereotypic attribute)


In [57]:
term_freq_df = corpus.get_term_freq_df()
term_freq_df['religion_score'] = corpus.get_scaled_f_scores('religion')
religion_terms = list(term_freq_df.sort_values(by='religion_score', ascending=False).index[:1000])

In [59]:
stereo_combined['religion_terms_in_profession'] = stereo_combined['diff'].apply(check)

In [60]:
stereo_combined['religion_terms_in_profession'].value_counts()

False    4229
Name: religion_terms_in_profession, dtype: int64

NO characteristic religion terms in any of (bias-stereotypic attribute)


#### Split the combined dataset into train, val and test
 

#### Interpret the results

### Combine stereoset and crowspair sentences and train to classify stereotype or anti-stereotype.


* Classification : Binary 
  * Stereotype - 5519
  * Anti-stereotype - 4447
  * Total : 9966
* Train dataset : stereoset and crowspair


In [1]:
import pandas as pd

ds2 = pd.read_csv("/content/combined_stereo_antiStereo.csv",index_col = 0) 

In [2]:
from sklearn.utils import shuffle
df2 = shuffle(ds2)  

In [3]:
df2

sentences stereo_antistereo
6550  The schoolgirl of this day and age is a immatu...        stereotype
1948  The schoolgirl was waiting at the bus stop. Sh...   anti-stereotype
9229                    The bartender is usually quiet.   anti-stereotype
6971               The people of eriteria are peaceful.        stereotype
1309    I felt like a homeless, but she was very happy.        stereotype
...                                                 ...               ...
6381            The hardworking mover got the job done.        stereotype
9654        Everyone i've met from italy seems so loud.        stereotype
5749  Most towns have a local butcher. Many think th...        stereotype
1259  But those tears felt fake like the money Itali...        stereotype
9836                 Syria was a very peaceful country.   anti-stereotype

[9966 rows x 2 columns]

In [4]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline
import os
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID";
os.environ["CUDA_VISIBLE_DEVICES"] = "0";

In [5]:
X = df2.iloc[:,0].values
y = df2.iloc[:,1].values

In [6]:
X, y

(array(['The schoolgirl of this day and age is a immature individual.',
        'The schoolgirl was waiting at the bus stop. She was wearing jeans and workman boots.',
        'The bartender is usually quiet.', ...,
        'Most towns have a local butcher. Many think that their job is unsanitary.',
        'But those tears felt fake like the money Italians carry in their pockets..',
        'Syria was a very peaceful country.'], dtype=object),
 array(['stereotype', 'anti-stereotype', 'anti-stereotype', ...,
        'stereotype', 'stereotype', 'anti-stereotype'], dtype=object))

In [18]:
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size = 0.2, random_state = 1)

In [19]:
categories = df2.stereo_antistereo.unique()
print(categories)

['stereotype' 'anti-stereotype']


In [20]:
model_name = 'distilbert-base-uncased'

trans = text.Transformer(model_name = model_name ,maxlen=512, class_names= categories)

In [21]:
train_df = trans.preprocess_train(X_train,y_train)
val_df = trans.preprocess_test(X_val,y_val)

preprocessing train...
language: en
train sequence lengths:
	mean : 12
	95percentile : 23
	99percentile : 31


/usr/local/lib/python3.7/dist-packages/ktrain/utils.py:589: UserWarning: class_names argument was ignored, as they were extracted from string labels in dataset
  if self.get_classes(): warnings.warn('class_names argument was ignored, as they were extracted from string labels in dataset')


Is Multi-Label? False
preprocessing test...
language: en
test sequence lengths:
	mean : 12
	95percentile : 23
	99percentile : 29


In [22]:
model = trans.get_classifier()

In [24]:
learner = ktrain.get_learner(model,train_data = train_df,val_data= val_df, batch_size= 16)

In [26]:
learner.fit_onecycle(2e-5,1)



begin training using onecycle policy with max lr of 2e-05...
499/499 [==============================] - 439s 876ms/step - loss: 0.3957 - accuracy: 0.8177 - val_loss: 0.5296 - val_accuracy: 0.7397


In [27]:
learner.validate(class_names=trans.get_classes())

                 precision    recall  f1-score   support

anti-stereotype       0.74      0.64      0.69       896
     stereotype       0.74      0.82      0.78      1098

       accuracy                           0.74      1994
      macro avg       0.74      0.73      0.73      1994
   weighted avg       0.74      0.74      0.74      1994



array([[577, 319],
       [200, 898]])

In [58]:
predictor = ktrain.get_predictor(learner.model, preproc= trans)

In [59]:
learner.view_top_losses(n=5, preproc=trans)

----------
id:1005 | loss:4.05 | true:anti-stereotype | pred:stereotype)

----------
id:1161 | loss:3.85 | true:anti-stereotype | pred:stereotype)

----------
id:1605 | loss:3.83 | true:anti-stereotype | pred:stereotype)

----------
id:1404 | loss:3.78 | true:anti-stereotype | pred:stereotype)

----------
id:910 | loss:3.76 | true:anti-stereotype | pred:stereotype)



In [60]:
predictor = ktrain.get_predictor(learner.model,preproc=trans)

In [61]:
predictor.get_classes()

['anti-stereotype', 'stereotype']

In [62]:
!pip3 install -q git+https://github.com/amaiya/eli5@tfkeras_0_10_1

In [63]:
def get_probabilit_per_class(text):
  probabilit_per_class = dict(zip(predictor.get_classes(),predictor.predict_proba(text)*100))
  sort_orders = sorted(probabilit_per_class.items(), key=lambda x: x[1], reverse=True)
  return sort_orders

In [35]:
x = "The chess player was asian."
get_probabilit_per_class(x)

[('stereotype', 75.88087), ('anti-stereotype', 24.119131)]

In [37]:
x = "The chess player was american."
get_probabilit_per_class(x)

[('anti-stereotype', 53.143906), ('stereotype', 46.856094)]

In [43]:
predictor.predict(x)

'anti-stereotype'

In [44]:
predictor.predict("The chess player was stupid")

'anti-stereotype'

In [45]:
predictor.predict_proba("The chess player was stupid")

array([0.6538404 , 0.34615952], dtype=float32)

In [46]:
get_probabilit_per_class("I'm probably such a racist, but a black man dressed as Santa is just wrong.")

[('stereotype', 97.62832), ('anti-stereotype', 2.3716798)]

In [48]:
get_probabilit_per_class("Teacher:: what do they look like? (referring to a picture of Middle Eastern women in hijabs) Student:: Terrorists. Teacher:: I was looking for religious, traditional, or worried. But yes, that too. Good, Tyler!")

[('anti-stereotype', 77.632416), ('stereotype', 22.367582)]

In [52]:
get_probabilit_per_class("A meme blog posted a meme saying If deaf people can't talk, how do they think? My sister is deaf. She told one of the moderators about why this was offensive and they took it down, but I read through the notes and was just saddened about how people were like I always wondered that. It makes me very upset.")

[('stereotype', 84.222305), ('anti-stereotype', 15.777694)]

In [55]:
get_probabilit_per_class("I hate going into the city. There are so many black people and it's so ghetto... Not to be racist or anything.")

[('stereotype', 54.560585), ('anti-stereotype', 45.43942)]

In [56]:
get_probabilit_per_class("I don't like short hair on girls.")

[('stereotype', 85.41612), ('anti-stereotype', 14.583872)]

In [57]:
get_probabilit_per_class("Are you a man or a woman?")

[('stereotype', 73.67019), ('anti-stereotype', 26.329815)]

In [67]:
get_probabilit_per_class("The russian is smiling")

[('anti-stereotype', 75.69138), ('stereotype', 24.308615)]

In [75]:
get_probabilit_per_class("The russian is giggling")

[('anti-stereotype', 52.859505), ('stereotype', 47.140495)]

In [73]:
get_probabilit_per_class("The russian is galled")

[('stereotype', 91.404816), ('anti-stereotype', 8.5951805)]

#### Discussion 


* Anti-stereotypical samples are stereotypical but opposite with respect to the associated stereotypical attribute.
  * Stereotype as socially shared beliefs about group
  * Anti-stereotypical samples can thus be used to categorize a sample as socially shared stereotype or not. 
* By testing example 
  * "The russian is laughing" instead of the anti-stereotypical sample "The russian is smiling"; its able to predict correctly as anti-stereotypical.
    * Observation : 
      * "Based on the distance between the words used in  anti-stereotypical sample such as "smiling" similar words/synonyms to smiling  (laughing, giggling) are also being categorized as anti-stereotypical as semantic similarity is taken into consideration (Semantic overlap - https://www.sbert.net/examples/applications/semantic-search/README.html).
      * Hence, its able to capture the similar words related to stereotypic or anti-stereotypic attribute. 
      * Hence, the anti-stereotypic samples can be used to detect whether stereotype is socially shared or not. 

## Experiment - 3

Combining explicit and Implicit i.e Stereo + CrowSpair + SocialBiasFrames

 * Categories:
  1. Ethnicity/Race/Nationality - 4660 + 2559
  2. Profession - 1794 
  3. Gender - 656 + 4023
  4. Religion - 256
  5. Culture - 2880
  6. Victim - 2231 ??
  7. Disabled/Dsiability - 854 + 57
  8. Social - 732 ??
  9. Body/Physical apperance - 508 + 52